lib/OpenQA/Worker/Cache/Client.pm

```diff
@@ -0,0 +1,291 @@
+use OpenQA::Worker::Common;
+
+use OpenQA::Worker::Cache::Request;
+use OpenQA::Worker::Cache::Request::Asset;
+use OpenQA::Worker::Cache::Request::Sync;
+
+use Mojo::Base -base;
+use Mojo::URL;
+use Mojo::File 'path';
+
+has host  => 'http://127.0.0.1:7844';
+has retry => 5;
+has cache_dir =>
+  sub { $ENV{CACHE_DIR} || (OpenQA::Worker::Common::read_worker_config(undef, undef))[0]->{CACHEDIRECTORY} };
+has ua => sub { Mojo::UserAgent->new };
+
+sub _url { Mojo::URL->new(shift->host)->path(shift)->to_string }
+
+sub _status {
+    my $request = shift;
+    return !!0 unless my $st = $request->result->json->{status} // $request->result->json->{session_token};
+    return $st;
+}
+
+sub _result {
+    my ($self, $field, $request) = @_;
+    eval {
+        $self->ua->post($self->_url('status') => json => {lock => $request->lock, id => $request->minion_id})
+          ->result->json->{$field};
+    };
+}
+
+sub _query {
+    my ($self, $q) = @_;
+    return _status(_retry(sub { $self->ua->get($self->_url($q)) } => $self->retry));
+}
+
+sub _post {
+    my ($self, $q, $j) = @_;
+    return _status(_retry(sub { $self->ua->post($self->_url($q) => json => $j) } => $self->retry));
+}
+
+sub _retry {
+    my ($cb, $times) = @_;
+    my $res;
+    my $att = 0;
+    $times ||= 0;
+
+    do { ++$att and $res = $cb->() } until $res->success || $att >= $times;
+
+    return $res;
+}
+
+sub _reply { !!(($_[0] eq STATUS_ENQUEUED) || ($_[0] eq STATUS_DOWNLOADING) || ($_[0] eq STATUS_IGNORE)) }
+
+sub _dequeue_lock { !!(shift->_post(dequeue => {lock => pop}) eq STATUS_PROCESSED) }
+
+# See also OpenQA::Cache::Request, as it's being consumed in the following methods
+sub execute_task {
+    my ($self, $request) = @_;
+    my $response = _retry(
+        sub {
+            $self->ua->post($self->_url('execute_task') => json =>
+                  {task => $request->task, args => [$request->to_array], lock => $request->lock});
+        } => $self->retry
+    );
+    my $json = $response->result->json;
+    $request->minion_id($json->{id}) if exists $json->{id};
+    return _status($response);
+}
+
+sub status { $_[0]->_post(status => {lock => $_[1]->lock, id => $_[1]->minion_id}) }
+sub processed { !!(shift->status(shift) eq STATUS_PROCESSED) }
+sub output    { shift->_result(output => pop) }
+sub result    { shift->_result(result => pop) }
+sub info      { $_[0]->ua->get($_[0]->_url("info")) }
+sub available { shift->info->success }
+sub available_workers {
+    return unless $_[0]->available;
+    return unless my $res = $_[0]->info->result->json;
+    return !!($res->{active_workers} != 0 || $res->{inactive_workers} != 0);
+}
+sub session_token { shift->_query('session_token') }
+sub enqueue       { _reply(shift->execute_task(@_)) }
+
+# TODO: This could go in a separate object (?)
+sub asset_path { path(shift->cache_dir, @_ > 1 ? (OpenQA::Worker::Cache::_base_host($_[0]) || shift) : ())->child(pop) }
```

Comment:
Nope, I did not forgot, the note is left intentionally there, I believe this could be furthermore extracted, but it is beginning to become a bit outside of scope of this PR